Hash functions for public keys and scripts

dr-orlovsky opened the pull request:
Introduces functions to compute bitcoin hashes for `PublicKey` and `Script` which were strangely missed out

stevenroose commented on src/blockdata/script.rs:
```diff
@@ -216,6 +217,16 @@ impl Script {
 
+    /// Returns 160-bit hash of the script
+    pub fn script_hash(&self) -> ScriptHash {
+        ScriptHash::hash(&self.serialize())
```
This makes an unnecessary allocation. Please serialize directly into the hashengine.

apoelstra replied:
This one is actually an allocation, and needs to be addressed

apoelstra replied:
Never mind, it is fixed in a later commit.

stevenroose commented on src/blockdata/script.rs:
```diff
@@ -216,6 +217,16 @@ impl Script {
+
+    /// Returns 256-bit hash of the script for P2WSH outputs
+    pub fn wscript_hash(&self) -> WScriptHash {
+        WScriptHash::hash(&self.serialize())
```
Idem.

stevenroose commented on src/util/key.rs:
```diff
@@ -80,6 +82,24 @@ pub struct PublicKey {
+    /// Returns bitcoin 160-bit hash of the public key
+    pub fn pubkey_hash(&self) -> PubkeyHash {
+        if self.compressed {
+            PubkeyHash::hash(&self.key.serialize())
```
Idem about allocation.

dr-orlovsky replied:
Sorry, after re-reviewing this I do not see where the allocation happens. The function serializes as `[u8; cost]`, hash function takes `&[u8]` and puts it right into the engine – so we save nothing by repeating the code from inside `hash` fn.

apoelstra replied:
Yeah, it's an unfortunate name for `key.serialize()`, since most every other serialization method returns a Vec..

stevenroose commented on src/util/key.rs:
```diff
@@ -80,6 +82,24 @@ pub struct PublicKey {
+    /// Returns bitcoin 160-bit hash of the public key for witness program
+    pub fn wpubkey_hash(&self) -> WPubkeyHash {
+        if self.compressed {
+            WPubkeyHash::hash(&self.key.serialize())
```
Idem about allocation.

dr-orlovsky commented:
@stevenroose I am not happy with all hash-related stuff still. Hashes are used everywhere, and even after hash type refactoring there is a lot of duplicated code (https://github.com/rust-bitcoin/rust-bitcoin/pull/391#discussion_r366243967) or unclear hash generation conventions (`txid`, `wtxid`, `bitcoin_hash`, we got the rid of, `block_hash`). I would prefer to have a compatible APIs for Pubkeys/Scripts, so if we generate a hash from the data structure with `*_hash` method, why we need to change it now? Or if we change it, isn't it that the rest of hash generations have to be refactored, to have `Txid::from_tx(&tx)` instead of `tx.txid()`?

stevenroose commented:
Hmm, yeah I agree semantically it makes more sense to have `<type>.<hash-type>()` methods.

I don't think the engine-based hash generation needs a shortcut. It's actually pretty common to do that in many other languages: you create a hasher (engine), write data in (we use `consensus_encode`) and then finalize the hasher (`Hash::from_engine`).

elichai commented:
> I don't think the engine-based hash generation needs a shortcut. It's actually pretty common to do that in many other languages: you create a hasher (engine), write data in (we use `consensus_encode`) and then finalize the hasher (`Hash::from_engine`).

I'd actually even say that most languages have *only* a hasher by default (except obviously python) even the  rust `sha2` crate by default shows to use the engine

dr-orlovsky commented:
I think I can explain the pattern with the Engine. It's kind of Java-style which have appeared from Enterprise patterns and used for doing some work that require internal state, costly at initialization. So instead of initializing all internal state time after time for each operation, a class is preferred, that stores that pre-initialized state and does operation-specific tasks in form of instance functions. However, this is not the case if you initialize such engine each time, so in our situation from best practice perspective, we either have to add a class-level method for engine to do the hashing task, or work on the persistence of pre-initialized `HashEngine` instances with sort of singleton pattern or such.

elichai commented:
> I think I can explain the pattern with the Engine. It's kind of Java-style which have appeared from Enterprise patterns and used for doing some work that require internal state, costly at initialization. So instead of initializing all internal state time after time for each operation, a class is preferred, that stores that pre-initialized state and does operation-specific tasks in form of instance functions. However, this is not the case if you initialize such engine each time, so in our situation from best practice perspective, we either have to add a class-level method for engine to do the hashing task, or work on the persistence of pre-initialized `HashEngine` instances with sort of singleton pattern or such.

The reason hashes usually come in engines is different.
it meant such that you don't need to store all the data in a buffer to hash it, you can stream it into the engine and whenever the "block" is full it will run the internal hash function and continue streaming, that way you can ie easily hash a 1GB file without doing any heap allocation or something like that (taking minimal memory)

dr-orlovsky commented:
However bitcoin hashes operate with a very small-sized data, "streamed" in 95% cases at once, so they do not benefit from this pattern at all. And when this is the case (like Taproot tagged hashes), the midstate is stored in a different way and embedded into the type itself...

stevenroose commented:
I think even for small values, we can benefit from this pattern. There is by definition 0 overhead of using an engine over using the `HashTrait::hash` function, because that one uses an engine internally. So the difference is the allocation of the data when using the latter method.

So even for very small bits of data, where allocation might be small, it's still more efficient. But then there is the problem that even for small values, the compiler nor the code can estimate the size of the allocation. So in many cases there will be frequent re-allocations. 

The two main things that are hashed are block headers and transactions. Headers are always 80 bytes, but the compiler doesn't know that and our code also doesn't anticipate that. So here it depends on the default allocation size of `Vec<u8>` to take enough space whether it will be 1 or multiple allocations.

For transactions, it becomes a lot more complex. There is really no way for the compiler, nor the serialization code, to know the size of a serialized tx. Transactions can range in size from a ~100 bytes to many kilobytes. While serializing them into a `Vec<u8>`, the vector will be frequently re-allocated to make more space.

I really think the usage of the hash engines you mention is both standard practice and not a problem for readability or concise-ness. It's 2 lines of boilerplate instead of 1.

stevenroose commented:
So I would suggest to reduce this PR to just adding those hash methods where they are missing, using the same pattern used elsewhere.
```rust
let mut engine = HashYouWant::engine();
whatever_data_to_be_hashes.consensus_encode(&mut engine).expect("engines don't error");
more_data_to_be_hashes.consensus_encode(&mut engine).expect("engines don't error");
HashYouWant::from_engine(engine)
```

dr-orlovsky commented:
Maybe it worth adding something like
```rust
macro_rules! hash {
    ($hash:ident, $($item:expr),+) => {
        {
            let mut engine = $hash::engine();
            $(
                $item.consensus_encode(&mut engine).expect("engines don't error");
            )+
            $hash::from_engine(engine)
        }
    }
}
```

stevenroose commented:
While that looks nice in theory, in many cases it won't work because also other data is pushed into the hash engine. And I feel like it's better to always do the same thing instead of have a macro that hides what's actually going on in half of the cases.

With `rg -trust "::engine()" -A 10` I found 7 cases where we could apply your macro, and 8 cases where we couldn't:

- `blockdata/block.rs`:
```rust
109:        let mut encoder = WitnessCommitment::engine();
110-        witness_root.consensus_encode(&mut encoder).unwrap();
111-        encoder.input(witness_reserved_value);
112-        WitnessCommitment::from_engine(encoder)
```
- `util/bip32.rs`:
```rust
566:        let mut engine = XpubIdentifier::engine();
567-        self.public_key.write_into(&mut engine);
568-        XpubIdentifier::from_engine(engine)
```
- `util/bip143.rs`:
```rust
49:            let mut enc = SigHash::engine();
50-            for txin in &tx.input {
51-                txin.previous_output.consensus_encode(&mut enc).unwrap();
52-            }
53-            SigHash::from_engine(enc)
--
57:            let mut enc = SigHash::engine();
58-            for txin in &tx.input {
59-                txin.sequence.consensus_encode(&mut enc).unwrap();
60-            }
61-            SigHash::from_engine(enc)
--
65:            let mut enc = SigHash::engine();
66-            for txout in &tx.output {
67-                txout.consensus_encode(&mut enc).unwrap();
68-            }
69-            SigHash::from_engine(enc)
```
- `util/address.rs`:
```rust
246:        let mut hash_engine = PubkeyHash::engine();
247-        pk.write_into(&mut hash_engine);
--
268:        let mut hash_engine = WPubkeyHash::engine();
269-        pk.write_into(&mut hash_engine);
--
283:        let mut hash_engine = WPubkeyHash::engine();
284-        pk.write_into(&mut hash_engine);
```

dr-orlovsky commented:
True. Ok, agree with you approach then. Sorry for taking that long to discuss: I'd like to understand rationale for design decisions in rust-bitcoin better (to follow it in other projects on top of it), that's why keep digging :)

dr-orlovsky commented:
We have to decide on closing this issue. While there is not a lot of use cases for the macro in `rust-bitcoin` itself, in the library we are building for L2/L3 there is a lot; so it's either have to go into `tust-bitcoin` or be kept on top of it. What would you suggest?

The reference implementation is here: https://github.com/pandoracore/rust-bitcoin/blob/ddb3efff95d8b39ce7a6fee0514c18d1a9c785b6/src/hash_types.rs#L84

If you are up for it I'll add it to this PR instead of the current code. Otherwise I will close the request.

apoelstra commented:
My feeling is that this functionality belongs in miniscript. I'm also not thrilled about doing things to support p2sh, which I feel is insecure (only 80 bits collision resistant), and is less efficient and more complicated than p2wsh in all cases.

dr-orlovsky commented:
I try to move it there, will share WIP

dr-orlovsky commented:
Rebased and fixed serializaiton (now it's non-allocating one, @stevenroose)

@apoelstra pls let me know if you still think it belongs to miniscript. I was going to transfer it there, but it's impossible to extend external rust types, and the function looks very native for `(W)ScriptHash` and `(W)PubkeyHash`; moreover they are quire similar by semantics to `Transaction.(w)txid()`, `Block.block_hash()` and other functions already residing in rust-bitcoin.

dr-orlovsky commented:
Update: @stevenroose Sorry, after re-reviewing this I do not see where the allocation happens in directly calling `hash` when a single `&[u8]` is given, like in `(w)pubkey_hash()` - pls see my comment in the code. As for script serialization, it's true that we need to use `script.as_bytes()` instead of `script.serialize()` (this saves from vec allocation), but again, it's safe to use `hash` and not duplicate the code with the engine.

apoelstra reviewed (dismissed):
Looks good! I skipped over the discussion about hash engines vs single-line functions, as it seems like it's moot now.

dr-orlovsky commented:
@stevenroose can you pls look at this PR so we can close it?
